typescript/sdk/src/core/HyperlaneRelayer.ts

```diff
@@ -0,0 +1,221 @@
+  ): Promise<DerivedIsmConfig> {
+    const config = (this.cache?.ism[chain]?.[ism] ??
+      (await new EvmIsmReader(this.multiProvider, chain).deriveIsmConfig(
+        ism,
+      ))) as DerivedIsmConfig | undefined;
+
+    if (!config) {
+      throw new Error(`ISM config not found for ${ism}`);
+    }
+
+    if (this.cache) {
+      this.cache.ism[chain] ??= {};
+      this.cache.ism[chain][ism] = config;
+    }
+
+    return config;
+  }
+
+  async getSenderHookConfig(
+    message: DispatchedMessage,
+  ): Promise<DerivedHookConfig> {
+    const originChain = this.core.getOrigin(message);
+    const hook = await this.core.getSenderHookAddress(message);
+    return this.getHookConfig(originChain, hook);
+  }
+
+  async getRecipientIsmConfig(
+    message: DispatchedMessage,
+  ): Promise<DerivedIsmConfig> {
+    const destinationChain = this.core.getDestination(message);
+    const ism = await this.core.getRecipientIsmAddress(message);
+    return this.getIsmConfig(destinationChain, ism);
+  }
+
+  async relayMessage(
+    dispatchTx: TransactionReceipt,
+    messageIndex = 0,
+    message = HyperlaneCore.getDispatchedMessages(dispatchTx)[messageIndex],
+  ): Promise<ethers.ContractReceipt> {
+    this.logger.info(`Preparing to relay message ${message.id}`);
+
+    const isDelivered = await this.core.isDelivered(message);
+    if (isDelivered) {
+      this.logger.debug(`Message ${message.id} already delivered`);
+      return this.core.getProcessedReceipt(message);
+    }
+
+    this.logger.debug({ message }, `Simulating recipient message handling`);
+    await this.core.estimateHandle(message);
+
+    // parallelizable because configs are on different chains
+    const [ism, hook] = await Promise.all([
+      this.getRecipientIsmConfig(message),
+      this.getSenderHookConfig(message),
+    ]);
+    this.logger.debug({ ism, hook }, `Retrieved ISM and hook configs`);
+
+    const metadata = await pollAsync(
+      () => this.metadataBuilder.build({ message, ism, hook, dispatchTx }),
+      5 * 1000, // every 5 seconds
+      12, // 12 attempts
+    );
+
+    this.logger.info({ message, metadata }, `Relaying message ${message.id}`);
+    return this.core.deliver(message, metadata);
+  }
+
+  async relayMessages(
+    dispatchTx: TransactionReceipt,
+  ): Promise<ethers.ContractReceipt[]> {
+    const messages = HyperlaneCore.getDispatchedMessages(dispatchTx);
+    return Promise.all(
+      messages.map((message, index) =>
+        this.relayMessage(dispatchTx, index, message),
+      ),
+    );
+  }
+
+  async relayMessageId(
+    originChain: ChainName,
+    messageId: string,
+  ): Promise<ethers.ContractReceipt> {
+    const dispatchTx = await this.core.getDispatchTx(originChain, messageId);
+    return this.relayMessage(dispatchTx);
+  }
+
+  hydrate(cache: RelayerCache): void {
+    assert(this.cache, 'Caching not enabled');
+    this.cache = objMerge(this.cache, cache);
+  }
+
+  async hydrateDefaults(): Promise<void> {
```

Comment:
I get the intention but I think `init` would be a more intuitive name for these fns for a new dev to guess that these must be called before the relayer can be used

Reply:
its not necessary for them to be called, but speeds up metadata building when a message is being processed (skipping derivation)

Reply:
agree that the naming is confusing here and probably leaks too much "implementation" detail